adapters/repos/db/vector/flat/index.go

```diff
@@ -0,0 +1,524 @@
+			return err
+		}
+	}*/
+	return nil
+}
+
+func byteSliceFromUint64Slice(x []uint64, slice []byte) []byte {
+	for i := range x {
+		binary.LittleEndian.PutUint64(slice[i*8:], x[i])
+	}
+	return slice
+}
+
+func uint64SliceFromByteSlice(x []byte, slice []uint64) []uint64 {
+	len := len(x) / 8
+	for i := 0; i < len; i++ {
+		slice[i] = binary.LittleEndian.Uint64(x[i*8:])
+	}
+	return slice
+}
+
+func (index *flat) Add(id uint64, vector []float32) error {
+	index.trackDimensionsOnce.Do(func() {
+		atomic.StoreInt32(&index.dims, int32(len(vector)))
+		index.bq = ssdhelpers.NewBinaryQuantizer()
+		index.pool = newPools()
+		if index.compression == flatent.CompressionNone {
+			return
+		}
+		atomic.StoreInt32(&index.dims, int32(len(vector)))
+		index.bq = ssdhelpers.NewBinaryQuantizer()
+	})
+	if index.compression == flatent.CompressionNone {
+		return nil
+	}
+	if len(vector) != int(index.dims) {
+		return errors.Errorf("insert called with a vector of the wrong size")
+	}
+	if index.distancerProvider.Type() == "cosine-dot" {
+		// cosine-dot requires normalized vectors, as the dot product and cosine
+		// similarity are only identical if the vector is normalized
+		vector = distancer.Normalize(vector)
+	}
+	vec, err := index.bq.Encode(vector)
+	if err != nil {
+		return err
+	}
+	slice := make([]byte, len(vec)*8)
+	index.storeCompressedVector(id, byteSliceFromUint64Slice(vec, slice))
+
+	return nil
+}
+
+func (index *flat) Delete(ids ...uint64) error {
+	return nil
+}
+
+func (index *flat) searchTimeEF(k int) int {
+	if index.compression == flatent.CompressionNone {
+		return k
+	}
+	// load atomically, so we can get away with concurrent updates of the
+	// userconfig without having to set a lock each time we try to read - which
+	// can be so common that it would cause considerable overhead
+	ef := int(atomic.LoadInt64(&index.ef))
+	if ef < k {
+		ef = k
+	}
+	return ef
+}
+
+func (index *flat) SearchByVector(vector []float32, k int, allow helpers.AllowList) ([]uint64, []float32, error) {
+	if index.distancerProvider.Type() == "cosine-dot" {
+		// cosine-dot requires normalized vectors, as the dot product and cosine
+		// similarity are only identical if the vector is normalized
+		vector = distancer.Normalize(vector)
+	}
+
+	ef := index.searchTimeEF(k)
+	heap := index.pqResults.GetMax(ef)
+	defer index.pqResults.Put(heap)
+
+	firstId := uint64(0)
+	alreadyFound := 0
+	if allow != nil {
+		firstId, _ = allow.Iterator().Next()
+	}
+
+	if index.compression != flatent.CompressionNone {
+		query, err := index.bq.Encode(vector)
+		if err != nil {
+			return nil, nil, err
+		}
+		cursor := index.store.Bucket(helpers.CompressedObjectsBucketLSM).Cursor()
+		var key []byte
+		var v []byte
+		if allow != nil {
+			buff := make([]byte, 16)
+			binary.BigEndian.PutUint64(buff[8:], firstId)
+			key, v = cursor.Seek(buff)
+		} else {
+			key, v = cursor.First()
+		}
+		for key != nil && (allow == nil || alreadyFound < allow.Len()) {
```

Comment:
I believe underlying bitmap performs some computation to get number of elements, meaning allowlist as well when `Len` is called. Since allowlist is not expected to be changed at that point `Len` can be assigned to variable.

Reply:
In case of allowlist contains values that are not present in the store, `alreadyFound >= allow.Len()` will never happen and cursor will walk until the end.
There can be additional break condition created (would require extending allowlist with method `Max()`, fairly easy with underlying bitmap): whenever key gets > max loop can be stopped as no more keys will match allowlist